File: library/src/androidTest/java/com/raizlabs/android/dbflow/test/querymodel/QueryModelTest.java
package com.raizlabs.android.dbflow.test.querymodel;

import android.database.Cursor;

import com.raizlabs.android.dbflow.sql.language.Delete;
import com.raizlabs.android.dbflow.sql.language.Select;
import com.raizlabs.android.dbflow.test.FlowTestCase;

import java.util.UUID;

/**
 * Description: Tests the {@link TestQueryModel} to ensure it works as expected.
 */
public class QueryModelTest extends FlowTestCase {


    public void testSalaryModel() {
        Delete.tables(SalaryModel.class);

        SalaryModel salaryModel = new SalaryModel();
        salaryModel.uid = UUID.randomUUID().toString();
        salaryModel.salary = 15000;
        salaryModel.name = "Andrew Grosner";
        salaryModel.department = "Developer";
        salaryModel.save();

        salaryModel = new SalaryModel();
        salaryModel.uid = UUID.randomUUID().toString();
        salaryModel.salary = 30000;
        salaryModel.name = "Bill Gates";
        salaryModel.department = "Developer";
        salaryModel.save();

        Cursor cursor = new Select(SalaryModel$Table.DEPARTMENT)
                .rawColumns("`SALARY` as average_salary", "`name` as newName")
                .from(SalaryModel.class).where().limit(1).groupBy(SalaryModel$Table.DEPARTMENT).query();

        TestQueryModel testQueryModel = new TestQueryModel();
        testQueryModel.loadFromCursor(cursor);

        assertTrue(testQueryModel.average_salary > 0);

        assertNotNull(testQueryModel.newName);

        assertNotNull(testQueryModel.department);
        assertEquals(testQueryModel.department, "Developer");

        Delete.tables(SalaryModel.class);
    }
}

File: dbflow-tests/src/test/java/com/raizlabs/android/dbflow/models/QueryModelTest.kt
package com.raizlabs.android.dbflow.models

import com.raizlabs.android.dbflow.BaseUnitTest
import com.raizlabs.android.dbflow.kotlinextensions.eq
import com.raizlabs.android.dbflow.kotlinextensions.exists
import com.raizlabs.android.dbflow.kotlinextensions.from
import com.raizlabs.android.dbflow.kotlinextensions.innerJoin
import com.raizlabs.android.dbflow.kotlinextensions.on
import com.raizlabs.android.dbflow.kotlinextensions.save
import com.raizlabs.android.dbflow.models.Author_Table.id
import com.raizlabs.android.dbflow.models.Blog_Table.author_id
import com.raizlabs.android.dbflow.models.Blog_Table.name
import com.raizlabs.android.dbflow.sql.language.SQLite.select
import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * Description: Tests to ensure we can load a Query model from the DB
 */
class QueryModelTest : BaseUnitTest() {


    @Test
    fun testCanLoadAuthorBlogs() {
        val author = Author(0, "Andrew", "Grosner")
        author.save()
        val blog = Blog(0, "My First Blog", author)
        blog.save()

        assert(author.exists())
        assert(blog.exists())

        val result = (select(name.withTable().`as`("blogName"), id.withTable().`as`("authorId"),
            Blog_Table.id.withTable().`as`("blogId")) from Blog::class innerJoin
            Author::class on (author_id.withTable() eq id.withTable()))
            .queryCustomSingle(AuthorNameQuery::class.java)!!
        assertEquals(author.id, result.authorId)
        assertEquals(blog.id, result.blogId)
    }
}
